### PR TITLE
Replace boost with C++11 [DISCL-391]

### PR DESCRIPTION
**Pull request by rdumusc**
Notes:
- FramePtr must remain a boost::shared_ptr as long as Tide is
  serializing it and does not mandate boost >= 1.56.
- boost::signals2 for Stream::disconnected is hard to replace,
  however it is only used internally by QmlStreamerImpl.cpp.
- program_options and unit_test could be optional.

### CODE REVIEW

File: deflect/Stream.h [hunk 1]
```diff
@@ -132,7 +126,7 @@ class Stream
     /** @name Asynchronous send API */
     //@{
     /** Future signaling success of asyncSend(). @version 1.1 */
```
Comment (dnachbaur) [inline deflect/Stream.h]: That breaks API imo. No longer version 1.1

**Review by dnachbaur — COMMENTED** [review]
Hmm, not removing boost entirely makes a bit wonder how useful that change is. But +1.

**Comment by rdumusc** [comment]
Yes, I thought I could easily remove all when I started. But I couldn't find an elegant solution for the shared_ptr serialization. I could copy the Frame in Tide from std::shared_ptr to boost::shared_ptr before serializing, which wouldn't be too expensive because the segments are QByteArray (implicitly shared) but still look a bit awkward. The boost::signal could be replaced by a callback function which would be fine for our use case. I was not sure and decided to stop here, which is still a step in the right direction (the only problem is the one API breakage you pointed out).

**Comment by rdumusc** [comment]
And I notice now that the API change requires a change in Equalizer because eq/deflect/Proxy.cpp because it explicitly uses boost in a make_ready_future() function...

**Comment by rdumusc** [comment]
Updated: I decided that since this was already breaking the API I might as well do all the changes and make boost optional.

File: CMakeLists.txt [hunk 2]
```diff
@@ -17,15 +17,13 @@ set(DEFLECT_LICENSE BSD)
 set(DEFLECT_DEPENDENT_LIBRARIES Boost)
```
Comment (dnachbaur) [inline CMakeLists.txt]: can be removed now?

File: doc/Changelog.md [hunk 3]
```diff
@@ -5,6 +5,9 @@ Changelog {#Changelog}
 
 ### 0.12.0 (git master)
 
+* [130](https://github.com/BlueBrain/Deflect/pull/130)
+  Replaced boost by C++11. Boost is now an optional dependency and it is used
+  only by the tests. Minor API changes were introdued by this change.
```
Comment (dnachbaur) [inline doc/Changelog.md]: Not so minor imo.
typo 'introdued'

File: CMakeLists.txt (hunk 2 above)
Comment (rdumusc) [reply CMakeLists.txt]: done

File: doc/Changelog.md (hunk 3 above)
Comment (rdumusc) [reply doc/Changelog.md]: done, replaced minor by some

**Comment by dnachbaur** [comment]
+1

**Pull request merged**